File: __test__/transformer/quantized-kv-cache.test.ts
/**
 * Tests for QuantizedKVCache functionality
 *
 * Tests verify:
 * 1. Basic creation and configuration
 * 2. Quantization/dequantization correctness
 * 3. Memory reduction compared to full precision
 * 4. Update and fetch operations
 * 5. Both 4-bit and 8-bit modes
 */

import { describe, it, expect } from 'vite-plus/test';
import { QuantizedKVCache, KVCache, MxArray } from '@mlx-node/core';

describe('QuantizedKVCache', () => {
  describe('Creation and Configuration', () => {
    it('should create with default config (8-bit, group_size=64)', () => {
      const cache = new QuantizedKVCache();
      expect(cache.bits).toBe(8);
      expect(cache.groupSize).toBe(64);
      expect(cache.offset).toBe(0);
    });

    it('should create 8-bit cache with custom config', () => {
      const cache = new QuantizedKVCache({ bits: 8, groupSize: 64 });
      expect(cache.bits).toBe(8);
      expect(cache.groupSize).toBe(64);
    });

    it('should create 4-bit cache', () => {
      const cache = new QuantizedKVCache({ bits: 4, groupSize: 32 });
      expect(cache.bits).toBe(4);
      expect(cache.groupSize).toBe(32);
    });

    it('should default to 8-bit for invalid bits value', () => {
      // 16 is invalid for quantized cache (use regular KVCache for 16-bit)
      const cache = new QuantizedKVCache({ bits: 16 });
      expect(cache.bits).toBe(8);
    });

    it('should reset cache correctly', () => {
      const cache = new QuantizedKVCache();
      cache.reset();
      expect(cache.offset).toBe(0);
    });
  });

  describe('Memory Usage', () => {
    it('should report zero memory for empty cache', () => {
      const cache = new QuantizedKVCache();
      expect(cache.memoryUsage()).toBe(0);
    });
  });

  describe('Update and Fetch Operations', () => {
    // Test dimensions
    const batch = 1;
    const nKvHeads = 2;
    const seqLen = 8;
    const headDim = 64; // Must be divisible by group_size

    it('should update and fetch with 8-bit quantization', () => {
      // Create test tensors: [batch, n_kv_heads, seq_len, head_dim]
      const keys = MxArray.randomUniform(BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)), -1.0, 1.0);
      const values = MxArray.randomUniform(
        BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)),
        -1.0,
        1.0,
      );

      const cache = new QuantizedKVCache({ bits: 8, groupSize: 64 });

      try {
        const [cachedKeys, cachedValues] = cache.updateAndFetch(keys, values);

        // Verify shapes
        const kShape = cachedKeys.shape();
        const vShape = cachedValues.shape();

        expect(kShape.length).toBe(4);
        expect(vShape.length).toBe(4);
        expect(Number(kShape[0])).toBe(batch);
        expect(Number(kShape[1])).toBe(nKvHeads);
        expect(Number(kShape[2])).toBe(seqLen);
        expect(Number(kShape[3])).toBe(headDim);

        // Verify offset updated
        expect(cache.offset).toBe(seqLen);

        // Verify memory usage increased
        expect(cache.memoryUsage()).toBeGreaterThan(0);
      } catch (e) {
        // Quantization may not be supported on all systems
        console.log('Skipping test - quantization may not be supported:', e);
      }
    });

    it('should update and fetch with 4-bit quantization', () => {
      const keys = MxArray.randomUniform(BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)), -1.0, 1.0);
      const values = MxArray.randomUniform(
        BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)),
        -1.0,
        1.0,
      );

      const cache = new QuantizedKVCache({ bits: 4, groupSize: 64 });

      try {
        const [cachedKeys, cachedValues] = cache.updateAndFetch(keys, values);

        // Verify shapes preserved
        const kShape = cachedKeys.shape();
        expect(Number(kShape[2])).toBe(seqLen);
        expect(cache.offset).toBe(seqLen);
      } catch (e) {
        console.log('Skipping 4-bit test - may not be supported:', e);
      }
    });

    it('should accumulate multiple updates correctly', () => {
      const cache = new QuantizedKVCache({ bits: 8, groupSize: 64 });

      try {
        // First update (prefill)
        const keys1 = MxArray.randomUniform(
          BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)),
          -1.0,
          1.0,
        );
        const values1 = MxArray.randomUniform(
          BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)),
          -1.0,
          1.0,
        );
        cache.updateAndFetch(keys1, values1);
        expect(cache.offset).toBe(seqLen);

        // Second update (decode step - single token)
        const keys2 = MxArray.randomUniform(BigInt64Array.from([batch, nKvHeads, 1, headDim].map(BigInt)), -1.0, 1.0);
        const values2 = MxArray.randomUniform(BigInt64Array.from([batch, nKvHeads, 1, headDim].map(BigInt)), -1.0, 1.0);
        const [cachedKeys, _] = cache.updateAndFetch(keys2, values2);

        // Should have accumulated
        expect(cache.offset).toBe(seqLen + 1);

        // Verify cached shape includes both updates
        const shape = cachedKeys.shape();
        expect(Number(shape[2])).toBe(seqLen + 1);
      } catch (e) {
        console.log('Skipping accumulation test:', e);
      }
    });

    it('should reset and allow reuse', () => {
      const cache = new QuantizedKVCache({ bits: 8, groupSize: 64 });

      try {
        // First sequence
        const keys = MxArray.randomUniform(
          BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)),
          -1.0,
          1.0,
        );
        const values = MxArray.randomUniform(
          BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)),
          -1.0,
          1.0,
        );
        cache.updateAndFetch(keys, values);
        expect(cache.offset).toBe(seqLen);

        // Reset
        cache.reset();
        expect(cache.offset).toBe(0);
        expect(cache.memoryUsage()).toBe(0);

        // New sequence
        cache.updateAndFetch(keys, values);
        expect(cache.offset).toBe(seqLen);
      } catch (e) {
        console.log('Skipping reset test:', e);
      }
    });
  });

  describe('Memory Comparison', () => {
    it('should use less memory than full precision cache (8-bit)', () => {
      const batch = 1;
      const nKvHeads = 4;
      const seqLen = 128;
      const headDim = 64;

      const keys = MxArray.randomUniform(BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)), -1.0, 1.0);
      const values = MxArray.randomUniform(
        BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)),
        -1.0,
        1.0,
      );

      // Full precision cache
      const fullCache = new KVCache();
      fullCache.updateAndFetch(keys, values);

      // Quantized cache
      const quantCache = new QuantizedKVCache({ bits: 8, groupSize: 64 });

      try {
        quantCache.updateAndFetch(keys, values);

        // Get memory usage
        // Full precision would be: batch * heads * seq * head_dim * 2 (bytes for bf16) * 2 (K+V)
        const fullMemory = batch * nKvHeads * seqLen * headDim * 2 * 2; // Approximate
        const quantMemory = quantCache.memoryUsage();

        console.log(`Full precision memory: ~${fullMemory} bytes`);
        console.log(`Quantized memory: ${quantMemory} bytes`);

        // 8-bit should use roughly half the memory (plus some overhead for scales/biases)
        // We expect at least 30% reduction
        expect(quantMemory).toBeLessThan(fullMemory * 0.7);
      } catch (e) {
        console.log('Skipping memory comparison test:', e);
      }
    });

    it('should use even less memory with 4-bit quantization', () => {
      const batch = 1;
      const nKvHeads = 4;
      const seqLen = 128;
      const headDim = 64;

      const keys = MxArray.randomUniform(BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)), -1.0, 1.0);
      const values = MxArray.randomUniform(
        BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)),
        -1.0,
        1.0,
      );

      const quant8Cache = new QuantizedKVCache({ bits: 8, groupSize: 64 });
      const quant4Cache = new QuantizedKVCache({ bits: 4, groupSize: 64 });

      try {
        quant8Cache.updateAndFetch(keys, values);
        quant4Cache.updateAndFetch(keys, values);

        const mem8 = quant8Cache.memoryUsage();
        const mem4 = quant4Cache.memoryUsage();

        console.log(`8-bit memory: ${mem8} bytes`);
        console.log(`4-bit memory: ${mem4} bytes`);

        // 4-bit should use roughly half of 8-bit
        expect(mem4).toBeLessThan(mem8 * 0.7);
      } catch (e) {
        console.log('Skipping 4-bit memory test:', e);
      }
    });
  });

  describe('Quality Validation', () => {
    it('should preserve approximate values after quantize/dequantize (8-bit)', () => {
      const batch = 1;
      const nKvHeads = 2;
      const seqLen = 4;
      const headDim = 64;

      // Create deterministic test data
      const keysData = new Float32Array(batch * nKvHeads * seqLen * headDim);
      for (let i = 0; i < keysData.length; i++) {
        keysData[i] = Math.sin(i * 0.1) * 0.5; // Values in [-0.5, 0.5]
      }
      const keys = MxArray.fromFloat32(keysData, BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)));
      const values = MxArray.fromFloat32(
        keysData, // Same data for simplicity
        BigInt64Array.from([batch, nKvHeads, seqLen, headDim].map(BigInt)),
      );

      const cache = new QuantizedKVCache({ bits: 8, groupSize: 64 });

      try {
        const [cachedKeys, _] = cache.updateAndFetch(keys, values);

        // Get values back as float32
        const originalData = keys.toFloat32();
        const recoveredData = cachedKeys.toFloat32();

        // Check that values are approximately equal
        // 8-bit quantization should have error < 1% for normalized data
        let maxError = 0;
        for (let i = 0; i < Math.min(100, originalData.length); i++) {
          const error = Math.abs(originalData[i] - recoveredData[i]);
          maxError = Math.max(maxError, error);
        }

        console.log(`Max quantization error (8-bit): ${maxError}`);
        // Allow up to 5% error for 8-bit
        expect(maxError).toBeLessThan(0.05);
      } catch (e) {
        console.log('Skipping quality test:', e);
      }
    });
  });
});

File: crates/mlx-core/src/transformer/quantized_kv_cache_test.rs
//! Tests for QuantizedKVCache functionality
//!
//! These tests verify:
//! 1. Basic creation and configuration
//! 2. Quantization/dequantization correctness
//! 3. Memory reduction compared to full precision
//! 4. Update and fetch operations
//! 5. Both 4-bit and 8-bit modes

#[allow(unused_imports)]
use super::KVCache;
use super::quantized_kv_cache::{
    QuantizedKVCache, QuantizedKVCacheConfig, UnifiedKVCache, create_unified_caches,
};
use crate::array::MxArray;

#[cfg(test)]
mod tests {
    use super::*;

    // ========================================================================
    // Basic Creation Tests
    // ========================================================================

    #[test]
    fn test_quantized_cache_default_config() {
        let cache = QuantizedKVCache::new(None);
        assert_eq!(cache.get_bits(), 8);
        assert_eq!(cache.get_group_size(), 64);
        assert_eq!(cache.get_offset(), 0);
        assert!(cache.is_empty());
    }

    #[test]
    fn test_quantized_cache_8bit_config() {
        let cache = QuantizedKVCache::new(Some(QuantizedKVCacheConfig {
            bits: Some(8),
            group_size: Some(64),
            step: Some(256),
        }));
        assert_eq!(cache.get_bits(), 8);
        assert_eq!(cache.get_group_size(), 64);
    }

    #[test]
    fn test_quantized_cache_4bit_config() {
        let cache = QuantizedKVCache::new(Some(QuantizedKVCacheConfig {
            bits: Some(4),
            group_size: Some(32),
            step: Some(128),
        }));
        assert_eq!(cache.get_bits(), 4);
        assert_eq!(cache.get_group_size(), 32);
    }

    #[test]
    fn test_quantized_cache_invalid_bits_defaults_to_8() {
        let cache = QuantizedKVCache::new(Some(QuantizedKVCacheConfig {
            bits: Some(16), // Invalid - not 4 or 8
            group_size: None,
            step: None,
        }));
        assert_eq!(cache.get_bits(), 8); // Should default to 8
    }

    #[test]
    fn test_quantized_cache_reset() {
        let mut cache = QuantizedKVCache::new(None);
        cache.reset();
        assert_eq!(cache.get_offset(), 0);
        assert!(cache.is_empty());
    }

    // ========================================================================
    // UnifiedKVCache Tests
    // ========================================================================

    #[test]
    fn test_unified_cache_full_precision() {
        let cache = UnifiedKVCache::new(16, 64);
        assert!(!cache.is_quantized());
        assert_eq!(cache.get_bits(), 16);
        assert_eq!(cache.get_offset(), 0);
    }

    #[test]
    fn test_unified_cache_8bit() {
        let cache = UnifiedKVCache::new(8, 64);
        assert!(cache.is_quantized());
        assert_eq!(cache.get_bits(), 8);
    }

    #[test]
    fn test_unified_cache_4bit() {
        let cache = UnifiedKVCache::new(4, 32);
        assert!(cache.is_quantized());
        assert_eq!(cache.get_bits(), 4);
    }

    #[test]
    fn test_create_unified_caches() {
        let caches = create_unified_caches(12, 8, 64);
        assert_eq!(caches.len(), 12);
        for cache in &caches {
            assert!(cache.is_quantized());
            assert_eq!(cache.get_bits(), 8);
        }
    }

    #[test]
    fn test_create_unified_caches_full_precision() {
        let caches = create_unified_caches(6, 16, 64);
        assert_eq!(caches.len(), 6);
        for cache in &caches {
            assert!(!cache.is_quantized());
            assert_eq!(cache.get_bits(), 16);
        }
    }

    // ========================================================================
    // Update and Fetch Tests (require MLX GPU context)
    // ========================================================================

    #[test]
    fn test_quantized_cache_update_and_fetch_8bit() {
        // Create test K/V tensors
        // Shape: [batch=1, n_kv_heads=2, seq_len=8, head_dim=64]
        // Note: head_dim must be divisible by group_size
        let batch = 1i64;
        let n_kv_heads = 2i64;
        let seq_len = 8i64;
        let head_dim = 64i64; // Divisible by group_size=64

        // Create random K/V tensors
        let keys =
            MxArray::random_uniform(&[batch, n_kv_heads, seq_len, head_dim], -1.0, 1.0, None)
                .expect("Failed to create keys");
        let values =
            MxArray::random_uniform(&[batch, n_kv_heads, seq_len, head_dim], -1.0, 1.0, None)
                .expect("Failed to create values");

        // Create 8-bit quantized cache
        let mut cache = QuantizedKVCache::new(Some(QuantizedKVCacheConfig {
            bits: Some(8),
            group_size: Some(64),
            step: Some(256),
        }));

        // Update and fetch
        let result = cache.update_and_fetch(&keys, &values);

        // Skip test if quantization is not available on this system
        if result.is_err() {
            eprintln!(
                "Skipping quantization test - may not be supported: {:?}",
                result.err()
            );
            return;
        }

        let (cached_keys, cached_values) = result.unwrap();

        // Verify shapes match
        let k_shape = cached_keys.shape().expect("Failed to get keys shape");
        let v_shape = cached_values.shape().expect("Failed to get values shape");

        assert_eq!(k_shape.len(), 4);
        assert_eq!(v_shape.len(), 4);
        assert_eq!(k_shape[0], batch);
        assert_eq!(k_shape[1], n_kv_heads);
        assert_eq!(k_shape[2], seq_len);
        assert_eq!(k_shape[3], head_dim);

        // Verify offset updated
        assert_eq!(cache.get_offset(), seq_len as i32);
    }

    #[test]
    fn test_quantized_cache_multiple_updates() {
        let batch = 1i64;
        let n_kv_heads = 2i64;
        let head_dim = 64i64;

        // Create cache
        let mut cache = QuantizedKVCache::new(Some(QuantizedKVCacheConfig {
            bits: Some(8),
            group_size: Some(64),
            step: Some(256),
        }));

        // First update
        let keys1 = MxArray::random_uniform(&[batch, n_kv_heads, 8, head_dim], -1.0, 1.0, None)
            .expect("Failed to create keys1");
        let values1 = MxArray::random_uniform(&[batch, n_kv_heads, 8, head_dim], -1.0, 1.0, None)
            .expect("Failed to create values1");

        let result1 = cache.update_and_fetch(&keys1, &values1);
        if result1.is_err() {
            eprintln!("Skipping - quantization not supported");
            return;
        }

        assert_eq!(cache.get_offset(), 8);

        // Second update (decode step - single token)
        let keys2 = MxArray::random_uniform(&[batch, n_kv_heads, 1, head_dim], -1.0, 1.0, None)
            .expect("Failed to create keys2");
        let values2 = MxArray::random_uniform(&[batch, n_kv_heads, 1, head_dim], -1.0, 1.0, None)
            .expect("Failed to create values2");

        let result2 = cache.update_and_fetch(&keys2, &values2);
        assert!(result2.is_ok());
        assert_eq!(cache.get_offset(), 9);

        // Verify accumulated cache shape
        let (cached_keys, _) = result2.unwrap();
        let shape = cached_keys.shape().expect("Failed to get shape");
        assert_eq!(shape[2], 9); // Total sequence length
    }

    #[test]
    fn test_unified_cache_full_precision_update() {
        let batch = 1i64;
        let n_kv_heads = 2i64;
        let head_dim = 64i64;

        let keys = MxArray::random_uniform(&[batch, n_kv_heads, 8, head_dim], -1.0, 1.0, None)
            .expect("Failed to create keys");
        let values = MxArray::random_uniform(&[batch, n_kv_heads, 8, head_dim], -1.0, 1.0, None)
            .expect("Failed to create values");

        let mut cache = UnifiedKVCache::new(16, 64); // Full precision

        let result = cache.update_and_fetch(&keys, &values);
        assert!(result.is_ok());

        let (cached_keys, cached_values) = result.unwrap();
        let k_shape = cached_keys.shape().expect("Failed to get shape");
        let v_shape = cached_values.shape().expect("Failed to get shape");

        assert_eq!(k_shape[2], 8);
        assert_eq!(v_shape[2], 8);
        assert_eq!(cache.get_offset(), 8);
    }

    #[test]
    fn test_unified_cache_reset() {
        let mut cache = UnifiedKVCache::new(8, 64);
        assert_eq!(cache.get_offset(), 0);

        cache.reset();
        assert_eq!(cache.get_offset(), 0);

        let mut cache_fp = UnifiedKVCache::new(16, 64);
        cache_fp.reset();
        assert_eq!(cache_fp.get_offset(), 0);
    }

    // ========================================================================
    // Memory Usage Tests
    // ========================================================================

    #[test]
    fn test_memory_usage_empty_cache() {
        let cache = QuantizedKVCache::new(None);
        let usage = cache.memory_usage();
        assert_eq!(usage, 0.0); // Empty cache should use no memory
    }

    // Note: Detailed memory comparison tests would require actual tensor operations
    // and are better suited for integration tests in TypeScript
}
